src/JsonSchema/Constraints/BaseConstraint.php

```diff
@@ -36,23 +37,28 @@ public function __construct(Factory $factory = null)
         $this->factory = $factory ?: new Factory();
     }
 
-    public function addError(JsonPointer $path = null, $message, $constraint = '', array $more = null)
+    public function addError(JsonPointer $path = null, ConstraintError $constraint = null, array $more = array())
     {
         $error = array(
             'property' => $this->convertJsonPointerIntoPropertyPath($path ?: new JsonPointer('')),
             'pointer' => ltrim(strval($path ?: new JsonPointer('')), '#'),
-            'message' => $message,
-            'constraint' => $constraint,
+            'message' => ucfirst(vsprintf($constraint->getMessage(), array_map(function ($val) {
```

Comment:
Remember that our goal is NOT to do any localizing ourselves, not now, not ever. All I'm shooting for in this PR is to:

1. centralize the error codes and internal error messages so that someone who DOES want to localize has an easy reference
2. Make the arguments that would be passed to a localizing mechanism easily available ( `params` in the error object), already in the correct order and ready to go.
3. Typehint addError somehow that it is impossible for json-schema devs (not localizers) to misuse.

That's pretty much it. So how WOULD someone do localization? Minimally, like this:

```
foreach(  $validator->getErrors() as $error ) {
    $msgId = "msg_{$error->constraint}";
    // insert your favorite localization technology here
    echo $i18n->loc($msgId, $error['params'];
}
```

I guess I would like to hear some feedback from one of the guys that wanted localization to see what he has in mind; maybe we could find a way to add some sugar.